File: java/sort/array/InsertionSort.java
package sort.array;

public class InsertionSort {
    public void sort(int[] nums) {
        for (int i = 1; i < nums.length; ++i) {
            int pivot = nums[i];
            if (pivot >= nums[i - 1]) {
                continue;
            }

            int start = 0;
            int end = i - 1;
            while (start <= end) {
                int middle = start + ((end - start) >> 1);
                if (pivot < nums[middle]) {
                    end = middle - 1;
                } else {
                    start = middle + 1;
                }
            }

            for (int j = i - 1; j >= start; --j) {
                nums[j + 1] = nums[j];
            }
            nums[start] = pivot;
        }
    }
}

File: py/sort/array/InsertionSort.py
#!/usr/bin/env python
# -*- coding:gbk -*-

from typing import List

class InsertionSort:
    def sort(self, nums: List[int]) -> None:
        for i in range(1, len(nums)):
            pivot = nums[i]
            if pivot >= nums[i - 1]:
                continue

            start, end = 0, i - 1
            while start <= end:
                middle = start + ((end - start) >> 1)
                if pivot < nums[middle]:
                    end = middle - 1
                else:
                    start = middle + 1

            for j in range(i - 1, start - 1, -1):
                nums[j + 1] = nums[j]
            nums[start] = pivot
